### PR TITLE
BUGFIX: Avoid inheritance on unset Fusion pathes

### PR DESCRIPTION
**Pull request by cvette**
When Fusion pathes are unset with the `>` operator the operations was previously handled by the parser. The drawback was that the parser unset could not affect inherited pathes which were merged into the prototype by the runtime.

This change adds the internal-key `__stopInheritanceChain` to the fusion ast whenever a value is unset. The fusion runtime will stop merging parent prototypes for the pathes where this key is set.

As a common example the following code does not render an empty body but a will render the 
keys  `site` and `node` from the `Neos.Neos:Page` prototype which is totally unexpected and against the intention.
```
prototype(Vendor.Site:ExampleDocument) < prototype(Neos.Neos:Page) {
    body >
    body = Neos.Fusion:Array {
       @position = 'after bodyTag'
       message = "hello world"
   }
}
```

!!! While this is a clear bug this change might break code that relied on the old inconsitent behavior. It is also important that this also avoids inheriting of `@position` keys so if you rely on the ordering you have to add the @position keys again after unsetting like in the example above.

Fixes: #2213, #1109

### CODE REVIEW

**Comment by jonnitto** [comment]
Hi @cvette 
Thanks for your PR! As this is a breaking change, I would suggest adding this in Neos 5.0. Could you write some tests for this behavior?

**Comment by cvette** [comment]
There are still a few cases which this fix does not solve.

Related: #1109

**Comment by mficzel** [comment]
Would like to keep this open as this is an existing source of confusion

**Comment by mficzel** [comment]
I understand the problem and the test but have trouble understanding the implementation. Since this is the fusion core this partly is to be expected but this makes the review hard.

@cvette can you explain a bit why and how you solved it. Maybe we should slack or hangout for that. I would like to get this fixed as this often causes trouble.

**Comment by cvette** [comment]
I'll need some time to get into this again. As mentioned above, this does not solve all cases. Before proceeding, I would like to solve this for all cases and make sure this is an actual solution rather than a workaround. When I wrapped my head around this, I'd be happy to discuss this with you. 

I'm sure we can get this ready for Neos 5.0 :)

**Comment by mficzel** [comment]
@cvette would love to get this sorted, i run into this often when i override the body of Neos.Neos:Page.

BTW: Another solution i would be fine with would be to remove the > operator at all. I find it mostly confusing that this thing exists and behaves that strangely.

**Comment by cvette** [comment]
```
processors.newSyntax.unset {
	simple = 'Foo'
	simple.@process.1 = ${value + 'bar'}
	simple.@process.2 = ${value + 'baz'}

	simple.@process.1 >
}
```

Is this a use case we need to support? Doesn't make much sense to me.

**Comment by mficzel** [comment]
@cvette i think yes. Does not make much sense but I would expect this to work

**Comment by cvette** [comment]
:heavy_check_mark: Added some more test cases. 
:heavy_check_mark: Tests are ok (requires neos/flow-development-collection#1623).
:heavy_check_mark: Demo site works.

Why the current implementation doesn't work
---------------------------------------------------------
The parser simply sets the value in the object tree to null. When the runtime merges the configuration for a path, it is unable to determine if it should override the inherited path.

See the following example:
```
prototype(Foo) {
    foo >
    foo = Neos.Fusion:Value
}
```
In this case, any information about the ```foo >``` would be lost, because the parser immediately overrides it.

This is how it is fixed
--------------------------
1. When the **parser** encounters a ```>```, the path is reset in the object tree and an internal property ```__valueUnAssignment``` is added to that path.
2. When the **runtime** evaluates a path, it merges different parts of the object tree created by the parser. The merging is done by ``Arrays::arrayMergeRecursiveOverruleWithCallback``
3. This method is adjusted to allow conditionally resetting the first array value before merging the second array value. In this case, it is reset when the second array value contains the internal property ```__valueUnAssignment```.

**Comment by mficzel** [comment]
I still plan to look at this but have to find the time

File: Neos.Fusion/Classes/Core/Parser.php [hunk 1]
```diff
@@ -512,11 +512,13 @@ protected function parseValueAssignment($objectPath, $value)
      *
      * @param string $objectPath The object path as a string
      * @return void
+     * @throws Fusion\Exception
      */
     protected function parseValueUnAssignment($objectPath)
     {
         $objectPathArray = $this->getParsedObjectPath($objectPath);
         $this->setValueInObjectTree($objectPathArray, null);
+        $this->setValueInObjectTree($objectPathArray, ['__valueUnAssignment' => true]);
```
Comment (mficzel) [inline Neos.Fusion/Classes/Core/Parser.php]: The key `__valueUnAssignment ` is a bit confusing. How about a more explicit name like `__stopInheritenceChain`

File: Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+prototype(Baz) >
+prototype(Baz) < prototype(Neos.Fusion:Join) {
+  k2 = 'Bar'
+}
+
+prototype(Baz) {
+  k3 = 'Baz'
+}
+
+unsets {
```
Comment (mficzel) [inline Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion]: Please give the pathes a speaking name.

File: Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+prototype(Baz) {
+  k1 = 'Foo'
+}
+
+prototype(Baz) >
+prototype(Baz) < prototype(Neos.Fusion:Join) {
+  k2 = 'Bar'
+}
+
+prototype(Baz) {
```
Comment (mficzel) [inline Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion]: Please give the prototype a speaking name.

File: Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    k4 = 'Quux'
+  }
+}
+
+prototype(Baz) {
+  k1 = 'Foo'
+}
+
+prototype(Baz) >
+prototype(Baz) < prototype(Neos.Fusion:Join) {
```
Comment (mficzel) [inline Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion]: Please give the prototype a speaking name.

File: Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    k3 = 'Baz'
+    k4 = 'Quux'
+  }
+}
+
+prototype(Baz) {
+  k1 = 'Foo'
+}
+
+prototype(Baz) >
```
Comment (mficzel) [inline Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion]: Please give the prototype a speaking name.

File: Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+prototype(Foo) < prototype(Neos.Fusion:Join) {
+  foo = Neos.Fusion:Join {
+    k3 = 'Baz'
+    k4 = 'Quux'
+  }
+}
+
+prototype(Baz) {
```
Comment (mficzel) [inline Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion]: Please give the prototype a speaking name.

File: Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+
+prototype(Bar) < prototype(Neos.Fusion:Join) {
+  k1 = 'Foo'
+  k2 = 'Bar'
+}
+
+prototype(Foo) < prototype(Neos.Fusion:Join) {
+  foo = Neos.Fusion:Join {
```
Comment (mficzel) [inline Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion]: Please give the prototype a speaking name.

File: Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    1 = 'Baz'
+  }
+}
+
+prototype(Bar) < prototype(Neos.Fusion:Join) {
+  k1 = 'Foo'
+  k2 = 'Bar'
+}
+
+prototype(Foo) < prototype(Neos.Fusion:Join) {
```
Comment (mficzel) [inline Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion]: Please give the prototype a speaking name.

File: Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+prototype(Neos.Fusion:ValueUnset2) < prototype(Neos.Fusion:ValueUnset1) {
+  value >
+  value = Neos.Fusion:Join {
+    1 = 'Baz'
+  }
+}
+
+prototype(Bar) < prototype(Neos.Fusion:Join) {
```
Comment (mficzel) [inline Neos.Fusion/Tests/Functional/FusionObjects/Fixtures/Fusion/Unset.fusion]: Please give the prototype a speaking name.

File: Neos.Fusion/Classes/Core/Runtime.php [hunk 10]
```diff
@@ -644,6 +654,9 @@ protected function getConfigurationForPath($fusionPath)
             $this->configurationOnPathRuntimeCache[$pathUntilNow]['p'] = $currentPrototypeDefinitions;
         }
 
+        // Remove any leftover internal __valueUnAssignment keys
+        unset($configuration['__valueUnAssignment']);
```
Comment (mficzel) [inline Neos.Fusion/Classes/Core/Runtime.php]: I do not understand why this line is needed here. I really do not like that the runtime manipulating the fusion ast here. Is this really necessary.

**Comment by mficzel** [comment]
And sorry for the long time until i finally reviewed this. The fusion runtime requires a special state of mind i cannot get into that easy.

File: Neos.Fusion/Classes/Core/Runtime.php (hunk 10 above)
Comment (cvette) [reply Neos.Fusion/Classes/Core/Runtime.php]: Technically, this is not really necessary. I mainly added this to not pollute the content cache entry identifier values. If you unset something like the GlobalCacheIdentifiers prototype (like it is done in the tests). The internal "unassignment" key is added to every cache entry identifier. 

I moved it to the content cache, so it is only removed from the entry identifier values array when rendering the identifier. Is that better?

**Comment by skurfuerst** [comment]
I just checked the code, and it makes sense to me. Thank you so much @cvette and @mficzel for working on this :heart:

All the best,
Sebastian

File: Neos.Fusion/Tests/Functional/FusionObjects/UnsetTest.php [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * @test
+     * @dataProvider unsetExamples
+     */
+    public function unsetWorks($path, $expected)
+    {
+        $view = $this->buildView();
+        $view->setFusionPath($path);
+        $this->assertEquals($expected, $view->render());
```
Comment (daniellienert) [inline Neos.Fusion/Tests/Functional/FusionObjects/UnsetTest.php]:
```suggestion
        self::assertEquals($expected, $view->render());
```

**Pull request merged**